### PR TITLE
Readline.bind_key and Readline.done added

### PR DESCRIPTION
**Pull request by daphee**
Expanded C bindings for readline. Now includes setting and getting the variable "rl_done" and calling the function "rl_bind_key".

### CODE REVIEW

File: src/readline.cr [hunk 1]
```diff
@@ -48,6 +50,20 @@ module Readline
...
+
+    LibReadline.rl_bind_key(c.ord, f)
+  end
+
+  def done
+    LibReadline.rl_done
+  end
+
+  def done=(val : Int)
+    LibReadline.rl_done = val
```
Comment (jhass) [inline src/readline.cr]: valid values are `0` and `1` I guess? How about mapping it to `Bool` if so?

File: src/readline.cr [hunk 2]
```diff
@@ -48,6 +50,20 @@ module Readline
     LibReadline.rl_point
   end
 
+  def bind_key(c : Char, f : Int,Int -> Int)
```
Comment (jhass) [inline src/readline.cr]: Note the alias in line 3, here we're outside the `lib`, so `Int != LibReadline::Int`

File: src/readline.cr [hunk 3]
```diff
@@ -48,6 +50,20 @@ module Readline
...
+    return c.ord if !(0 <= c.ord <= 255)
```
Comment (jhass) [inline src/readline.cr]: Looks like an error condition? Maybe raise `ArgumentError`?

**Comment by asterite** [comment]
Thanks!

I read a bit of the [IRC talk](http://irclog.whitequark.org/crystal-lang/2015-08-26#13823989;) and yes, this would need to be written in a different way. I guess you'll need a global `Hash(Int, KeyBindingProc)` and a single non-closure KeyBindingProc that checks that based on the key invokes the associated proc with a hash lookup. When invoking `bind_key` you'd set this non-closure proc.

The `Signal` module is done like that too.

Also, maybe it's better to define:

```crystal
 def bind_key(c : Char, &f : KeyBindingProc)
   # ...
 end
```

so you can use it like:

```crystal
Readline.bind_key('a') do
  puts "Hi!"
end
```

In this form you can use the usual block syntax, and you can even omit the arguments if you want to.

**Comment by daphee** [comment]
Defined bind_key now like you proposed it.

Use one single Proc now for handling keybindings:

```crystal
 KeyBindingHandler = ->(count : LibReadline::Int, key : LibReadline::Int) do
    if (handlers = @@key_bind_handlers) && handlers[key.to_i32]?
      res = handlers[key].call(count.to_i32, key.to_i32)
      LibReadline::Int.cast(res)
    else
      LibReadline::Int.cast(1)
    end
  end
```

**Comment by asterite** [comment]
Thanks, looks good!

I'm thinking that we might eventually move readline (and curses) outside of the main repository, because maybe these aren't that "core".

**Pull request merged**